src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,337 @@
+        else:
+            self.index = load_dataset(self.dataset, with_embeddings=True, with_index=True, split=self.dataset_split)
+        self._index_initialize = True
+
+    def get_doc_dicts(self, doc_ids):
+        return [self.index[doc_ids[i].tolist()] for i in range(doc_ids.shape[0])]
+
+    def get_top_docs(self, query_vectors, n_docs=5):
+        _, docs = self.index.get_nearest_examples_batch(self.index_name, query_vectors, n_docs)
+        ids = [[int(i) for i in doc["id"]] for doc in docs]
+        vectors = [doc[self.index_name] for doc in docs]
+        return torch.tensor(ids), torch.tensor(vectors)
+
+
+class RagRetriever(object):
+    """
+    A distributed retriever built on top of the ``torch.distributed`` communication package. During training all workers
+    initalize their own instance of the retriever, however, only the main worker loads the index into memory. The index is stored
+    in cpu memory. The index will also work well in a non-distributed setup.
+
+    Args:
+        config (:class:`~transformers.RagConfig`):
+            The configuration of the RAG model this Retriever is used with. Contains parameters indicating which ``Index`` to build.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        assert (
+            config.retriever_type == "hf_retriever" or config.retriever_type == "legacy_retriever"
+        ), "invalid retirever type"
+
+        self.retriever = (
+            HFIndex(config.dataset, config.dataset_split, config.index_name, config.index_path)
+            if config.retriever_type == "hf_retriever"
+            else LegacyIndex(config.retrieval_vector_size, config.index_path, config.passages_path)
+        )
+        self.process_group = None
+        self.n_docs = config.n_docs
+        self.batch_size = config.retrieval_batch_size * torch.cuda.device_count()
+
+    def init_retrieval(self, distributed_port):
+        """
+        Retrirever initalization function, needs to be called from the training process. The function sets some common parameters
+        and environment variables. On top of that, (only) the main process in the process group loads the index into memory.
+
+        If this functin doesn't get called, we assume we're operating in a non-distributed environment and the index gets loaded
+        at first query.
+
+        Args:
+            distributed_port (:obj:`int`):
+                The port on which the main communication of the training run is carried out. We set the port for retrieval-related
+                communication as ``distributed_port + 1``.
+        """
+
+        logger.info("initializing retrieval")
+
+        # initializing a separate process group for retrievel as the default
+        # nccl backend doesn't support gather/scatter operations while gloo
+        # is too slow to replace nccl for the core gpu communication
+        if dist.is_initialized():
```

Comment:
Maybe we should have all the distributed setting logic inside the `examples/` code ?